### DISCUSSION TITLE
Ignore spec/test file when building the app

### DISCUSSION THREAD

**Discussion by goenning** [vercel/next.js#23959] [Ideas]
# Feature request

## Is your feature request related to a problem? Please describe.

Next currently exposes spec files as an API route

## Describe the solution you'd like

spec|test files should be ignored by Next.js when compiling the app.

## Describe alternatives you've considered

Right now I just ignore it, but that's being needless deployed to Vercel

## Additional context

<img width="608" alt="Screenshot 2020-08-16 at 19 45 09" src="https://user-images.githubusercontent.com/94755/90341714-b3865f00-dff9-11ea-82d1-1dc52d068c2f.png">

**Comment by davidcalhoun** [comment]
It seems that at the moment they expect all tests to be located in `__tests__` at the root, but I am like you and also want to co-locate the tests with their code.  It appears the only way to do this at the moment is to import a Webpack plugin and configure it to ignore files there.

Edit: just found [this example](https://nextjs.org/docs/api-reference/next.config.js/custom-webpack-config) in the docs.

Edit 2: see below for `pageExtensions` workaround which works pretty nicely.

**Comment by jcamden** [comment]
I can't find it at the moment (because the moment is after 4am), but I thought I came across an issue thread in which users concluded that there is no way to co-locate tests with Webpack 5 enabled. Other users, if I remember correctly, validated my xp that using an IgnorePlugin with Webpack 5 enabled does nothing, i.e.: 
```
const webpack = require('webpack');

module.exports = {
  future: {
    // very sadly, revert to webpack 4 in order for IgnorePlugin (below) to successfully ignore co-located tests
    webpack5: false,
  },
  webpack: (config) => {
    // console.log(config.target);
    config.plugins.push(
      new webpack.IgnorePlugin({
        resourceRegExp: /.test.(ts|tsx)$/,
      })
    );
    return config;
  },
```
Two questions:
1) Can anyone find and link to the existing issue I described?
2) Does Next 11's default support for Webpack 5 include the capacity to ignore co-located tests and snapshots?

Thanks :)

**Comment by davidcalhoun** [comment]
The best workaround I've found is to [add a config to pageExtensions](https://nextjs.org/docs/api-reference/next.config.js/custom-page-extensions) and move all pages from `foo.tsx` to `foo.page.tsx`!  This allows you to add other co-located files like `foo.test.tsx` which are ignored!

**Reply by jcamden** [reply]
Wow, excellent! Thank you!

**Comment by rkristelijn** [comment]
`npm install --save-dev ignore-loader`

```
// next.config.js
/** @type {import('next').NextConfig} */
module.exports = {
  reactStrictMode: true,
  webpack: (config, { dev, isServer }) => {
    // Exclude *.spec.ts(x) files from being compiled by Next.js
    config.module.rules.push({
      test: /\.spec\.tsx?$/,
      use: 'ignore-loader',
    });

    return config;
  },
};
```

**Comment by JoeRoddy** [comment]
I'm guessing this is newer, as none of the comments mention it, but Vercel have their own `.vercelignore` file now:

https://vercel.com/docs/deployments/vercel-ignore
